cortex_m: Add debug symbols for OpenOCD/GDB thread support

daniel-k opened the pull request:
GDB has the ability to inspect threads, but to use this feature with RIOT the GDB-server, in this case OpenOCD, needs to know about the thread structures and the stacking. This PR introduces 2 new symbols in order to provide support for GDB threads.

To use this feature you need my patched OpenOCD version found [here](https://github.com/daniel-k/openocd/tree/riot_patched). I'm working on getting these patches upstream to OpenOCD so that one day this will be a default feature.

I added the needed `-rtos auto` line to every board that has a `openocd.cfg` file, but only tested `stm32f4discovery` and `samr21-xpro` yet.

OpenOCD can only access symbols, no struct offsets or define macros, so I introduced `max_threads` to iterate over `sched_threads` and `_tcb_name_offset` that will only be available when compiled with `DEVELHELP` to access the thread name. Although both symbols are declared as `volatile` they will be optimized away, so that's why I added the hack in `sched_task_exit()`. I'm sure there's a better way, so please tell me :)

daniel-k commented:
I just realized that `cortex_m` might not be a good title as this will introduce these symbols for all platforms, although it is only used by OpenOCD, i.e. Cortex-M-

OlegHahm commented:
very nice!

OlegHahm commented:
Works like charm on the iotlab-m3 with gnrc_networking:
```gdb
(gdb) info threads
  Id   Target Id         Frame 
  7    Thread 7 (at86rfxx :  : Blocked receive) 0x08001716 in _msg_receive (m=0x20000ed0 <_nomac_stacks+868>, block=1) at /home/oleg/git/RIOT/core/msg.c:329
  6    Thread 6 (udp :  : Blocked receive) 0x08001716 in _msg_receive (m=0x20004714 <_stack+850>, block=1) at /home/oleg/git/RIOT/core/msg.c:329
  5    Thread 5 (ipv6 :  : Blocked receive) 0x08001716 in _msg_receive (m=0x20001304 <_stack+858>, block=1) at /home/oleg/git/RIOT/core/msg.c:329
  4    Thread 4 (6lo :  : Blocked receive) 0x08001716 in _msg_receive (m=0x20003f64 <_stack+842>, block=1) at /home/oleg/git/RIOT/core/msg.c:329
  3    Thread 3 (pktdump :  : Blocked receive) 0x08001716 in _msg_receive (m=0x20003b80 <_stack+1384>, block=1) at /home/oleg/git/RIOT/core/msg.c:329
  2    Thread 2 (main :  : Blocked mutex) mutex_lock (mutex=0x20000390 <_rx_mutex>) at /home/oleg/git/RIOT/core/mutex.c:53
* 1    Thread 1 (idle :  : Running) idle_thread (arg=<optimized out>) at /home/oleg/git/RIOT/core/kernel_init.c:74
```

daniel-k commented:
@OlegHahm 
If you experience problems with backtrace when switching thread, cherry-pick https://github.com/daniel-k/openocd/pull/1

gebart commented:
Wow!
Do you have a link to an upstream ticket for tracking the OpenOCD integration?

daniel-k commented:
>  Do you have a link to an upstream ticket for tracking the OpenOCD integration?

http://openocd.zylin.com/#/c/3000/, but seems they take more time to review contributions :P

daniel-k commented:
So what's your opinion on this? Waiting for the OpenOCD change will not make sense I would say. I can't estimate how long they'll take for reviewing my commit and in the end it has to match the RIOT "interface". So it would make sense to agree on this first and then maybe accelerate the OpenOCD change.

gebart commented:
will `$_TARGETNAME configure -rtos RIOT` cause any issues when using a non-patched openocd?

gebart commented on core/sched.c:
```diff
@@ -181,6 +190,12 @@ void sched_switch(uint16_t other_prio)
 NORETURN void sched_task_exit(void)
 {
+    /* Prevent optimization of these symbols */
+    max_threads = max_threads;
```
This will cause warnings from various static analyzers. Does adding `__attribute__((used))` help?

daniel-k replied:
@gebart

Doesn't work. It is caused by the linker garbage collection:
```bash
export CFLAGS_LINK  = -ffunction-sections -fdata-sections
export LINKFLAGS += -Wl,--gc-sections
```
So there doesn't seem to be a way to prevent optimization from within the compiler other than removing `-fdata-sections`. But that's no option.

Adding
```
export LINKFLAGS += -Wl,-u,max_threads
```
to `boards/Makefile.include.cortexm_common` keeps the variable from being optimized away, but that's not a nice solution either.

Another possible solution would be to define a new section in RAM (linker script) with `KEEP(this-section)` and moving the variable into that section.

I would rather like to have a solution where it's only marked next to variable in source, but there doesn't seem to be a way. The latter solution is not far away from that at least.

daniel-k replied:
@gebart 
See e3a2dcf for how this could look like. This way these symbols should be optimizied away if a platform doesn't have the `KEEP` directive in the linkerscript, so no overhead for other platforms. Haven't tested though, I'm lacking the toolchains for msp430 and avr.

gebart commented:
I think still attribute((used)) will be necessary when building with LTO.
After a quick googling it seems like the only solution without actually using the symbols is to put them in a special section, like you did.

daniel-k commented:
> will `$_TARGETNAME configure -rtos RIOT` cause any issues when using a non-patched openocd?

In general this does no harm. I switched from `-rtos auto` to `-rtos RIOT` because the former didn't work out most of the time. Now I found out why: there was a bug in another RTOS that prevented RIOT from being detected (see [bugfix](http://openocd.zylin.com/#/c/3017/). In the end it makes no difference for us which to use now, I also updated my fork on Github, so if you pull and rebuild you'll have that, too.

daniel-k commented:
To get the OpenOCD patches reviewed I was told to send some RIOT guys over to them to review my patches, as you are really using RIOT. So please head over to [OpenOCD RIOT support](http://openocd.zylin.com/#/c/3000/) and maybe [OpenOCD mqx bugfix](http://openocd.zylin.com/#/c/3017/) to review my patches, i.e. try out and say it works.

You need to register at [their Gerrit](http://openocd.zylin.com/login/%23%2Fregister%2Fq%2Fstatus%3Aopen) but you can use your Github account via OAuth.

gebart commented:
This PR needs a rebase after tcb_t -> thread_t rename #4578 

OpenOCD side needs an update because it expects thread status to be of `uint16_t` type, but it is now `uint8_t` since #4924, all threads are otherwise listed as "Unknown Status"

daniel-k commented:
@gebart 
Thanks for the update. I don't have a board at hand so I can't test. I also updated my openocd fork on Github (not pushed upstream though): https://github.com/daniel-k/openocd/commit/01dc81399c45d3c8de4f31ff8535f6e0503357d7

OlegHahm commented:
How about moving this common line in some common config that gets sourced by all the different  boards?

OlegHahm commented:
I can confirm that this PR doesn't break anything with an unpatched openocd version, but with https://github.com/daniel-k/openocd/commit/01dc81399c45d3c8de4f31ff8535f6e0503357d7 I get `:3333: Connection timed out.` on samr21-xpro.

OlegHahm commented:
Problem was on my end - still works like charm.

OlegHahm commented:
The question is: should we merge it here or wait for the review on the OpenOCD site?

kYc0o commented:
Feature freeze -> postponed.

miri64 commented:
Postponed due to feature freeze.

OlegHahm commented:
> The question is: should we merge it here or wait for the review on the OpenOCD site?

Anyone with an opinion here? I would vote for merging this PR ASAP. At least one could use this nice feature on RIOT with a locally patched OpenOCD version.

daniel-k commented:
I'm currently not using RIOT in any project, but when working with it this was a crucial feature for me. Concerning the OpenOCD part: they are preparing the 0.10 release right now so I guess it's to late to get RIOT-support merged into 0.10 anymore. But still, when it's merged here it's certainly possible to get the patch accepted for the next release.

gebart commented:
I agree, this should be merged. Might need updates to some boards which were added after last rebase though.

OlegHahm commented:
@daniel-k, you could cherry-pick https://github.com/OlegHahm/RIOT/commit/6c22e3f77fed06349fa8c699233a66adbbd1bba8 to add the missing new boards.

daniel-k commented:
@OlegHahm cherry-picked it is. Thanks!

miri64 commented:
If I understood this correctly you need a patched version of openocd to have this feature, but without the patch it is still working (just tested it with `Open On-Chip Debugger 0.9.0 (2017-01-10-16:08)`). So I wouldn't be opposed to merging this either.

miri64 commented:
(But I leave the actual ACK to a person more knowledgeable with OpenOCD)

OlegHahm commented:
I'm without hardware right now. Can anyone just briefly check with any board using OpenOCD for flashing and an unpatched version, in order to assure that it is indeed not breaking anything? If the result is positive, I'm willing to ACK.

kYc0o commented:
I just tested this PR and it didn't break anything, however it seems to succeed always, even if it didn't.

OlegHahm commented:
> however it seems to succeed always, even if it didn't.

What do you mean by "succeed"?

kYc0o commented:
> What do you mean by "succeed"?

When the flashing was successful.

In current master it returns an error which stops make immediately. With this change openocd returns success even if the flashing failed. It's related to what we had before #2897.

OlegHahm commented:
This sounds odd. Have you locally merged this PR into master or just checked it out?

kYc0o commented:
Just checked it out locally, with a rebase it works ok.

daniel-k commented:
Finally \o/ Thanks guys :)

OlegHahm commented:
Well, thank you for the endurance. ;)

biboc commented:
It seems that openocd PR has never been merged
I'll try to push it to get merged
Debugging thread may be very useful

kfessel commented:
$_TARGETNAME configure -rtos auto

may leads seemigly to detection of a different rtos with Ubuntu upstream openocd-0.10.0-6

gdb> attach <openocd pid>

provides following backtrace

Thread 1 "openocd" received signal SIGSEGV, Segmentation fault.
uCOS_III_update_thread_offsets (rtos=0x564ae6dc72e0) at src/rtos/uCOS-III.c:193
193     src/rtos/uCOS-III.c: No such file or directory.

after that openocd dies
